### PR TITLE
Migration to TS2.0

### LINKED ISSUE #1: TypeScript 2.0 for dojo/widgets

### PR DESCRIPTION
**Pull request by agubler**
Dependant on has, shim, core & compose TS2.0 updates

 * [x] Upgrade to TS2.0 (beta)
 * [x] Convert to use `paths` and `baseUrl` for UMD type discovery
 * [ ] Add `strictNullChecks`
 * [x] Add `noImplicitThis `

Fixes #1

### CODE REVIEW

**Comment by agubler** [comment]
I'm going to say that perhaps this should go in without the `strictNullChecks` compiler option and create a task to complete this

File: src/createDijit.ts [hunk 1]
```diff
@@ -166,7 +165,7 @@ function resolveCtor<D extends DijitWidget>(Ctor: DijitWidgetConstructor<D> | st
 				handle.remove();
 				reject(error);
 			});
-			const mid = Ctor;
+			const mid = <string> Ctor;
```
Comment (kitsonk) [inline src/createDijit.ts]: Hmmm... I am surprised with flow control that this hadn't been narrowed to `string` here.

**Comment by kitsonk** [comment]
outside of the one comment, which I don't think we have to address right now 👍

File: src/createDijit.ts (hunk 1 above)
Comment (agubler) [inline src/createDijit.ts]: Lol, so the `next` TS version `v2.1.0-dev.20160815` doesn't narrow the type correctly but the version `v2.1.0-dev.20160817` does.

So I updated this and it worked locally as it used the `20160817` version but for some reason travis used `20160815` version and failed.

Comment (kitsonk) [inline src/createDijit.ts]: As mentioned in other channels `node_modules` is cached on Travis.  So once you have a branch that has cached, you would have to clear it to get the same behaviour.

**Pull request merged**